### ISSUE TITLE
allow retrieval of the list of public snapshots of a private projects

### ISSUE THREAD

**Issue by rentzso** [poldracklab/bids-core#2]
```
/api/projects/< project ID >/snapshots
```

This endpoint to get a list of snapshots for a given project appears to rely on the permissions of the project itself. Within our use case projects always remain private and only snapshots can be made public.  

So if a user is viewing a public snapshot they can't get a direct list of sibling snapshots that are also public. If a user makes this request to get the list of snapshots of a project that is private but has public snapshots they currently get the following response.

```
{
    "code": 403,
    "uid": null,
    "detail": "this request is not allowed"
}
```

we will return all public snapshots for some project even for a public request.

**Comment by rentzso** [comment]
@constellates I have pushed a new change. Let me know if the permissions logic is good for you.
Please note that authenticated requests will hide public snapshots by default if you don't use the flag "public".

**Comment by constellates** [comment]
I'm not sure I understand. I'm making authenticated requests without a public flag and I'm seeing both public and private snapshots.

```
 /api/projects/56c4f63e23a2430e64c64c74/snapshots
```

This behavior seems ok though (unless I'm misunderstanding). 

I did notice another permissions behavior that I think should change. Permissions for getting this list of snapshots or retrieving a single snapshot are being checked against the permissions on the snapshot itself. I think it would make more sense if they were checked against the parent project permissions, since the snapshots are immutable. This would prevent weird cases where new project members are unable to see existing snapshots.

I just sent @chrisfilo and email to check if this is correct before moving forward and am waiting to hear back.

**Comment by constellates** [comment]
So I heard back. The functionality I described is correct. We would like to be able to have a setup where if a user has permission to view a project they have permission to view the all the snapshots of that dataset regardless of the permissions on that snapshot when it was taken. 

I think the best solution for us would be, what I described above, to change the permissions check when retrieving snapshots (and lists) to check the permissions of the parent project instead of the permissions on the snapshots themselves.

One alternative that Chris proposed is to allow permissions to be modified on snapshots so we can batch update them when a project's permissions change. This would add some bloat to our application but should work if there's a reason we can't or shouldn't do the first solution.

@rentzso let me know your thoughts on this issue and these solutions. And let me know if you prefer I break this out to a new github issue.

**Comment by rentzso** [comment]
Using the permissions on the parent project is fine for me.

Just a clarification on the resulting logic:
if a user is revoked permissions on the parent project, she will continue to have permissions on all its snapshots. Can she view them in the project list or only the public ones? From the discussion we were having, I guess only the public ones, but let me know if I am right.

**Comment by constellates** [comment]
Awesome.

If a user's permission is revoked from the parent project, she would only be able to view the public snapshots for that project. And if she somehow got the id of a private snapshot that she has permissions on from before her project permission was revoked, she would not be able to retrieve that snapshot and instead get a 403 error.

**Comment by rentzso** [comment]
I thought we were talking only about the list of snapshots for a project, not the permissions on a single snapshot.

This is problematic also because the original project could be deleted. If you need to modify access to a single snapshot from the parent project, I think it is inevitable that you modify permissions on all the snapshots.

**Comment by constellates** [comment]
Yea. Sorry this started with the list issue and I diverged to this related issue. I hadn't thought about the case when a project is deleted but snapshots remain. You're right the only way to handle this will be through managing permissions on snapshots themselves. 

And if that's the case I no longer see any point in changing the list to check parent project permissions if it's already checking the snapshot permissions.  I'll open a separate issue for modifying snapshot permissions.

**Comment by rentzso** [comment]
no worries. It is great to have the right logic figured out.

**Comment by constellates** [comment]
Closing. I believe the only remaining issues with this are addressed in #5 .
